### PR TITLE
16098 — selecting chat message and pressing any key focus the message instead of switch the focus to the composer

### PR DESCRIPTION
**Pull request by robertKozik**
### Details

This PR creates new document-scoped key listener which is connected with composer lifecycle. Listener will be mounted only when composer is rendered. It listens on any alphanumeric key event, and focuses the composer accordingly

Also there is added a fix for the ROUTES method `parseReportRouteParams` as it was not suited to the `Navigation.getActiveRoute` returend path - it don't have slash at the beggining.

the copy&paste listener is now document-scoped rather than connected to textInput and restricted in the same manner as new key listner
### Fixed Issues

$ https://github.com/Expensify/App/issues/16098
PROPOSAL: https://github.com/Expensify/App/issues/16098#issuecomment-1602615946

### Tests

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift)
4. Verify that the composer is focused again and the above character is entered in the composer
5. Open emoji picker
6. Press any character keys like a,b,c,d
7. Verify that the emoji composer is focused again and the above character is entered in this composer. And the main composer is not affected

As the test steps also the [regression tests](https://github.com/Expensify/App/pull/17649#issuecomment-1557577683) pasted by @PauloGasparSv in previous PR can be used :

### Typing should focus on the composer
1. Open any chat
2. Click on a message
3. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Verify that the composer is focused and the characters you type are inserted there

### Typing in other inputs and modals
(Coming from [here](https://github.com/Expensify/App/issues/17478))

1. Open any chat
2. Open the Global Search and make sure the search is working
3. Blur the global search input and start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Make sure the characters you typed aren’t added to the search input or the Chat composer box
5. Open your profile then display name and make sure the “First name” and “Last name” are not focused then start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
6. Go back to the chat and make sure those characters were not added to the composer box
7. Go to Settings > Workspaces
8. Open any workspace and go to the “Manage Members” page
9. With no input focused, start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…) and make sure they are not added to the Chat composer box
10. Click on “Invite” and repeat the previous step

### The composer Emoji Picker
(Coming from https://github.com/Expensify/App/issues/17413)

1. Open any chat
2. Click on the Emoji Picker icon in the right side of the composer box
3. Make sure you can search inside the Emoji Picker but what you type isn’t added to the composer box
4. With the Emoji Picker opened, blur the search box and start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
5. Make sure the focus goes back to the Emoji Picker search box and all characters are added there instead of the Chat composer box
6. Hover any message and click on the “Add reaction” icon
7. Repeat steps 3 to 5

### The FAB popover menu

1. Open any chat
2. Click on the Floating Action Button (`+`) in the bottom left of the screen
3. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Make sure that characters you typed aren't added to the Chat composer box
5. Close popover of the Floating Action Button, blur the chat composer if it's in focus
6. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
7. Make sure that characters you typed are added to the Chat composer box

### "Meeting" popup

1. Open any chat
2. Click on the headphone icon on the right of chat header
3. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Make sure that characters you typed aren't added to the Chat composer box
5. Close popover, blur the chat composer if it's in focus
6. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
7. Make sure that characters you typed are added to the Chat composer box

### three dot popup

1. Open any task
2. Click on the three dot icon on the right of chat header
3. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
4. Make sure that characters you typed aren't added to the Chat composer box
5. Close popover, blur the chat composer if it's in focus
6. Start typing alpha-numeric characters (like a,b,c,d… but not Enter, Shift…)
7. Make sure that characters you typed are added to the Chat composer box

### Copying&Pasting - Web & Desktop only
(Coming from https://github.com/Expensify/App/issues/17397 and https://github.com/Expensify/App/issues/17411)

1. Open any chat
8. Select/Highlight the text from a message
9. Copy the text using CMD + C or CTRL + C
10. Click on the composer and paste the text using CMD + V or CTRL + V

- [X] Verify that no errors appear in the JS console

### Offline tests

Same as test steps

### QA Steps

Same as test steps

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/61146594/46d7a227-4fda-4277-94f5-546b45b48cb6

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/61146594/a4a6e718-290c-42ef-87e7-ab5317cfa4f1

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/61146594/ecf229e3-b3b7-424c-94d1-9013aaf60041

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/61146594/2dac93f6-08b7-4b4a-9a21-3906e206bcdf

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/61146594/814f427b-3ed6-4413-8d2b-b07c795b92d1

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/61146594/8bfe814d-4a95-4500-873e-b0695f3aa265

</details>

### CODE REVIEW

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Tests mentioned here work well, please include the linked https://github.com/Expensify/App/pull/17649#issuecomment-1557577683 steps here as well, it would be easy for everyone!

**Comment by robertKozik** [comment]
@Santhosh-Sellavel Sure, pasted the testes 👌🏼

**Comment by Santhosh-Sellavel** [comment]
Bug: Pressing copy paste is broken

Steps:
1. Select a text in the message
2. Press CMD + C 
3. Manually focus the composer.
4. Press CMD + V

Data is pasted multiple times

#### Comparision between staging & dev
https://github.com/Expensify/App/assets/85645967/d144a3d3-9613-4918-b1ed-ce23585a74aa

**Comment by Santhosh-Sellavel** [comment]
Resolve conflicts and look into the above bug I will continue testing tomorrow!

**Comment by robertKozik** [comment]
@Santhosh-Sellavel @PauloGasparSv  I want to clarify one thing about this issue/PR. From this issue: https://github.com/Expensify/App/issues/17411 (linked in the repro steps) I assumed that copy&pasting function which will focus composer from anywhere should be implemented. But I missed that in this issue that was pointed out as `New Feature`. Thus should this behaviour be included in this PR?

**Comment by robertKozik** [comment]
With new changes from last commit everything should be fine. Looks like after redirecting to the new page, the composer is not unmounting. Because of it, I bind the lifecycle of the listeners with the navigation focus/blur, so we wont register multiple ones - which was causing the breakage.

**Comment by Santhosh-Sellavel** [comment]
@robertKozik 
Can re-test the PR now all tests are failing
 
Now the Typing does not focus on the composer

https://github.com/Expensify/App/assets/85645967/d15c4e50-53ea-41e4-b8d8-487888c36769

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Also fix the lint Screenshot 2023-06-30 at 4 44 46 AM

File: src/pages/home/report/ReportActionCompose.js [hunk 1]
```diff
@@ -52,6 +52,10 @@ import * as TaskUtils from '../../../libs/actions/Task';
 import * as Browser from '../../../libs/Browser';
 import * as IOU from '../../../libs/actions/IOU';
 import PressableWithFeedback from '../../../components/Pressable/PressableWithFeedback';
+import * as KeyDownListener from '../../../libs/KeyboardShortcut/KeyDownPressListener';
+import * as EmojiPickerActions from '../../../libs/actions/EmojiPickerAction';
+import Navigation from '../../../libs/Navigation/Navigation';
```
Comment (Santhosh-Sellavel) [inline src/pages/home/report/ReportActionCompose.js:57]: Lint

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]

**Comment by robertKozik** [comment]
> @robertKozik Can re-test the PR now all tests are failing
> 
> Now the Typing does not focus on the composer
> 
>  Screen.Recording.2023-06-30.at.4.42.11.AM.mov

@Santhosh-Sellavel Can you double check if you were using my branch? I checked it locally and It's working on my end

https://github.com/Expensify/App/assets/61146594/11fa0f02-5451-4db8-ad8e-34a2eaa7cf98

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Try merging with the latest main, seems it fails after merging please check!

**Comment by robertKozik** [comment]
@Santhosh-Sellavel Damn, you were right. EmojiPicker functional migration do a little of harm here, but now should be all right!

**Comment by Santhosh-Sellavel** [comment]
@robertKozik Still lint checks fail, can you check and update that

**Comment by Santhosh-Sellavel** [comment]
Bug: There is a space extra space is added after the first character.

Steps:
1. Select any message to unfocus the composer
2. Type abcd
3. Expected `abcd` is entered in composer
4. Actual `a bcd` is entered

https://github.com/Expensify/App/assets/85645967/d8831a97-4a7d-446f-a256-ecfc0dba15e9

### If enter the characters more fastly there is space between each character

https://github.com/Expensify/App/assets/85645967/fbb04ee4-dd75-4066-b81e-f86b0df8605b

cc: @robertKozik

File: src/pages/home/report/ReportActionCompose.js [hunk 2]
```diff
@@ -662,20 +679,53 @@ class ReportActionCompose extends React.Component {
...
-     * Callback for the emoji picker to add whatever emoji is chosen into the main input
-     *
-     * @param {String} emoji
+     * Callback to add whatever text is chosen into the main input (used f.e as callback for the emoji picker)
+     * @param {String} text
      */
-    addEmojiToTextBox(emoji) {
-        this.updateComment(ComposerUtils.insertText(this.comment, this.state.selection, `${emoji} `));
+    replaceSelectionWithText(text) {
+        this.updateComment(ComposerUtils.insertText(this.comment, this.state.selection, `${text} `));
```
Comment (Santhosh-Sellavel) [inline src/pages/home/report/ReportActionCompose.js:686]: Extra space is added from here, caution when modifying as we need space at the end for emoji but not for the text.

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
Mostly looks good, waiting to resolve the bug!

File: src/pages/home/report/ReportActionCompose.js (hunk 2 above)
Comment (robertKozik) [reply src/pages/home/report/ReportActionCompose.js]: Thanks for pointing this place out

**Comment by robertKozik** [comment]
@Santhosh-Sellavel lint and bug fixed 🚀

**Comment by PauloGasparSv** [comment]
@robertKozik we have new conflicts here, can you fix them you have the chance?

File: src/pages/home/report/ReportActionCompose.js [hunk 3]
```diff
@@ -209,8 +209,8 @@ class ReportActionCompose extends React.Component {
         // so we need to ensure that it is only updated after focus.
         const isMobileSafari = Browser.isMobileSafari();
 
-        this.unsubscribeBlur = () => null;
-        this.unsubscribeFocus = () => null;
+        this.unsubscribeNavBlur = () => null;
+        this.unsubscribeNavFocus = () => null;
```
Comment (Santhosh-Sellavel) [inline src/pages/home/report/ReportActionCompose.js:213]: Can we have the full name please, no short names

Comment (Santhosh-Sellavel) [reply src/pages/home/report/ReportActionCompose.js:213]: Seems it was not done here

**Comment by robertKozik** [comment]
I see checks failed after my merge, I’ll resolve them in short time

**Comment by robertKozik** [comment]
All good!

File: src/components/Composer/index.js [hunk 4]
```diff
@@ -159,8 +166,15 @@ class Composer extends React.Component {
...
         if (this.textInput) {
-            this.textInput.addEventListener('paste', this.handlePaste);
             this.textInput.addEventListener('wheel', this.handleWheel);
+
+            // we need to handle listeners on navigation focus/blur as Composer is not unmounting
+            // when navigating away to different report
+            this.unsubscribeFocus = this.props.navigation.addListener('focus', () => document.addEventListener('paste', this.handlePaste));
+            this.unsubscribeBlur = this.props.navigation.addListener('blur', () => document.removeEventListener('paste', this.handlePaste));
+
+            // focus is not triggered on component mount so we need to add listener manually
```
[the remark this answers is not in the document]
Comment (robertKozik) [reply src/components/Composer/index.js:176]: It refers to line 177, which calls the paste addEventListner independently from the focus listener. This has to be done this way, as mounting the component isn't connected with the navigation focus/blur cycle.
I'll rephrase this comment to be more clear though

[the remark this answers is not in the document]
Comment (robertKozik) [reply src/components/Composer/index.js]: changed, let me know if now it's clearer

[the remark this answers is not in the document]
Comment (PauloGasparSv) [reply src/components/Composer/index.js]: Thks so much!!!

**Comment by Santhosh-Sellavel** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web & Desktop </summary>

#### Copy/Paste Case

https://github.com/Expensify/App/assets/85645967/960ed844-8876-4f52-bf5c-adb786133a4d

#### Global modal & emoji case
https://github.com/Expensify/App/assets/85645967/d1a63f2e-8922-49d1-8647-da0c35a028f8

</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by Santhosh-Sellavel** [comment]
Bug: When Global Create is active composer still captures the text with being focused

https://github.com/Expensify/App/assets/85645967/8c8f0f12-b0d6-4234-b631-3c74ec8bf003

**Comment by PauloGasparSv** [comment]
Nice catch, thks for tagging me :D 

This looks buggy specially because as we type we keep on focusing on one of the actions, so I think we shouldn't allow typing while that is open

@robertKozik can you can include a fix for that + add test/qa steps for this too?

**Comment by robertKozik** [comment]
@PauloGasparSv Sure, I'm on it. I need to think how to detect the open FAB, as I don't see it's status being exported in any way.

**Comment by robertKozik** [comment]
@Santhosh-Sellavel @PauloGasparSv I've moved the ref for the `FloatingActionButtonWithPopoverMenu` to separate file in order to access it from `ReportActionCompose`. It's similar to the already existing implementation of `EmojiPickerActions` with which we are checking the emoji picker visibility status. New test case is also added to the original post - `The FAB popover menu`.

File: src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopoverUtils.js
Comment (Santhosh-Sellavel) [inline src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopoverUtils.js:1]: This should be moved to `libs` folder

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]

**Comment by robertKozik** [comment]
Suggestion Resolved

**Comment by Santhosh-Sellavel** [comment]
Sorry @robertKozik I missed this component earlier, can we have some other hook to detect visible popups?

https://github.com/Expensify/App/assets/85645967/17885596-ac2c-4bd9-b174-57ccaf594951

**Comment by robertKozik** [comment]
@Santhosh-Sellavel I've switched the way of detecting such a modals to use onyx data (`ONYXKEYS.MODAL`), it should be sufficient. Also I've updated test steps with the `headphone` popover and `three dots` popover inside task screen

**Comment by Santhosh-Sellavel** [comment]
Can you please resolve conflicts?

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM tests well!

**Review by PauloGasparSv — APPROVED** [review]
LGTM!

<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/6564265/7914579c-45b9-4245-a1c5-6588a81fff81

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Skipped, still having problems with chrome in the android emulator

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/6564265/8bdbdc83-cd2e-4b68-adb3-621ceecf49da

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/6564265/14e608a2-c72d-459e-9ad1-60daa64b353c

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/6564265/095e5b9d-c9b1-41f3-82f0-e535f452ebdd

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/6564265/a54bd7f6-ab32-40e3-971e-b1c479679614

</details>

File: src/pages/home/report/ReportActionCompose.js [hunk 5]
```diff
@@ -664,20 +681,56 @@ class ReportActionCompose extends React.Component {
...
+            return;
+        }
+
+        // if we're typing on another input/text area, do not focus
+        if (['INPUT', 'TEXTAREA'].includes(e.target.nodeName)) {
+            return;
+        }
+
+        this.focus();
+        this.replaceSelectionWithText(e.key, false);
```
Comment (situchan) [inline src/pages/home/report/ReportActionCompose.js:731]: This call was redundant. Later it caused https://github.com/Expensify/App/issues/33710.
More details about the root cause: https://github.com/Expensify/App/issues/33710#issuecomment-1871243110

**Pull request merged**